### PR TITLE
Add Support for Pass-Through Hosts

### PR DESCRIPTION
**Pull request by mjallday**
By adding a prefix to the path of the certificate like this:
- `ssl/pt/tnt8ciq1xog.sandbox.verygoodproxy.io/tenant1c.testclients.verygoodsecurity.io` (Note the `pt/` path)

It is possible to control the `Host` header that is passed through to the vault reverse proxy application sitting behind this server.

By doing this we have more control over the route that will be selected and can use this information to offer multiple reverse proxies to our customers.

### Example
- `vault.sbx.customer-domain-a.com -> tenancy.verygoodproxy.com -> api.sbx.customer-domain-a.com`
- `vault.sbx.customer-domain-b.me -> tenancy.verygoodproxy.com -> api.sbx.customer-deomain-b.me`
 
Both host names are passed to the reverse proxy. Before this change is applied the reverse proxy app would receive `tenancy.verygoodproxy.com` as the host name and match on that to choose the route to send the traffic.

With this change enabled for a single tenant the proxy will receive the host name requested by the user e.g. `vault.sbx.customer-domain-a.com` and be able to route based on that.

### CODE REVIEW

File: templates/vault_tenant.conf.j2 [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
 # This template represents a termination point for a SSL certificate
 
+{% set url_parts = item.split('/') %}
+{% set proxy_address = url_parts[-2] %}
+{% set custom_host_name = url_parts[-1] %}
+{% set passthrough_host_name = url_parts[1] == 'pt' %}
+
+{% if proxy_address.split('.') | length > 1 %}
+
+{% set data_environment = proxy_address.split('.')[1] %}
```
[the remark this answers is not in the document]
Comment (max-lobur) [reply templates/vault_tenant.conf.j2]: we can lookup tenant by tnt prefix, and data env by looking for live/sandbox -> get rid of indexing and folding assumptions

**Comment by mjallday** [comment]
@zdmytriv @max-lobur i think i've solved the issue you've highlighted. please take a look and let me know what you think. 

we could alternatively also encode it in the path structure.

**Comment by mjallday** [comment]
@akhon i tried re-working the playbook to use this non-standard dns entry. it really complicates the code and i think the better approach here is 

1. modify the path structure used for nginx-ssl-terminator
1. update the lendup ssl structure to use same format as the original dns setup. it should be `tnt123.live.proxy.lendupvault.com` instead of having `proxy` before `live`. `live` should be the first element after the `tenant123` identifier. 

i don't know how much work is involved in doing the second approach so we may be forced to the first route.

**Comment by mjallday** [comment]
@max-lobur can you take this from here?

I've added a script to migrate the certs

`AWS_PROFILE=vgs-dev AWS_DEFAULT_REGION=us-west-2 python migrate.py vault-dev-01-audits-01-citadel-ob362bzzrwee`. Add param `--no-dry-run` to make it work and actually copy. 

I've tested this and after it's done you can see the results

```
(nginx-ssl-router) ➜  nginx-ssl-router git:(pass-through) ✗ AWS_PROFILE=vgs-dev AWS_DEFAULT_REGION=us-west-2 aws s3 ls --recursive s3://vault-dev-01-audits-01-citadel-ob362bzzrwee/ssl/
2018-03-21 15:42:25       5189 ssl/2/np/reverse.proxy.live.vgs.lendupvault.com/live/tntbce09fb6/lu.testclients.verygoodsecurity.io
2018-03-21 15:42:25       5189 ssl/2/np/reverse.sandbox.verygoodproxy.io/sandbox/tnt8ciq1xog/tenant1a.testclients.verygoodsecurity.io
2018-03-21 15:42:25       5185 ssl/2/np/reverse.sandbox.verygoodproxy.io/sandbox/tnt8ciq1xog/tenant1b.testclients.verygoodsecurity.io
2018-03-21 15:42:25       5185 ssl/2/np/reverse.sandbox.verygoodproxy.io/sandbox/tntbxmlbl0g/tenant2.testclients.verygoodsecurity.io
2018-03-21 15:42:25       3516 ssl/2/np/reverse.sandbox.verygoodproxy.io/sandbox/tntmp6ltmwa/terminator-test.verygoodproxy.io
2017-10-30 15:41:06       5136 ssl/live.verygoodproxy.io
2017-10-30 15:41:16       5148 ssl/sandbox.verygoodproxy.io
2018-02-19 10:10:41       5189 ssl/tnt8ciq1xog.sandbox.verygoodproxy.io/tenant1a.testclients.verygoodsecurity.io
2018-02-19 14:48:40       5185 ssl/tnt8ciq1xog.sandbox.verygoodproxy.io/tenant1b.testclients.verygoodsecurity.io
2018-03-15 16:53:27       5189 ssl/tntbce09fb6.proxy.live.vgs.lendupvault.com/lu.testclients.verygoodsecurity.io
2018-02-19 10:11:28       5185 ssl/tntbxmlbl0g.sandbox.verygoodproxy.io/tenant2.testclients.verygoodsecurity.io
2017-10-30 07:40:25       3516 ssl/tntmp6ltmwa.sandbox.verygoodproxy.io/terminator-test.verygoodproxy.io
```

Once that works we can run it for real like `AWS_PROFILE=vault-prod AWS_DEFAULT_REGION=us-west-2 python migrate.py vault-production-01-audits-01-citadel-2q67wpt30wla` and add the `--no-dry-run` arg again.

@zdmytriv if we do this we'll need to make sure that the cert manager is aware of this new directory structure. is there any issue there?

**Comment by mjallday** [comment]
Example rendered from new directory structure.

```
# Ansible managed
# This template represents a termination point for a SSL certificate



server {

    listen              443 ssl proxy_protocol;
    server_name         tenant1a.testclients.verygoodsecurity.io;
    ssl_certificate     /etc/nginx/conf.d/tenant1a.testclients.verygoodsecurity.io.cert;
    ssl_certificate_key /etc/nginx/conf.d/tenant1a.testclients.verygoodsecurity.io.cert;
    ssl_protocols       TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers         HIGH:!aNULL:!MD5;

    set_real_ip_from 0.0.0.0/0;
    real_ip_header proxy_protocol;
    real_ip_recursive on;

    location / {
      proxy_set_header        Host reverse.sandbox.verygoodproxy.io;
      proxy_set_header        Via terminator;
      proxy_set_header        X-Forwarded-Host $host;
      proxy_set_header        X-Real-IP $proxy_protocol_addr;
      proxy_set_header        X-Forwarded-For $proxy_protocol_addr;
      proxy_set_header        X-Forwarded-Proto $scheme;
      proxy_set_header        VGS-Tenant tnt8ciq1xog;
      resolver                8.8.8.8;

      set                     $vault    https://reverse.sandbox.verygoodproxy.io;
      proxy_pass              $vault;
      proxy_read_timeout      90;
    }

    set $tenant tnt8ciq1xog;
    set $data_environment sandbox;
    set $tenant_endpoint $http_host;
    access_log /var/log/nginx/access.log json_combined;

}
```

File: templates/vault_tenant.conf.j2 [hunk 2]
```diff
@@ -1,33 +1,45 @@
...
+
 server {
+
     listen              443 ssl proxy_protocol;
-    server_name         {{ item | basename }};
-    ssl_certificate     {{ ssl_router_nginx_config_dir }}{{ item | basename }}.cert;
-    ssl_certificate_key {{ ssl_router_nginx_config_dir }}{{ item | basename }}.cert;
+    server_name         {{ custom_host_name }};
+    ssl_certificate     /etc/nginx/conf.d/{{ custom_host_name }}.cert;
+    ssl_certificate_key /etc/nginx/conf.d/{{ custom_host_name }}.cert;
```
Comment (mjallday) [inline templates/vault_tenant.conf.j2]: this needs to be reverted. i took this from the dockerize branch but this is incorrect for the non-dockerized version.

Comment (mjallday) [reply templates/vault_tenant.conf.j2]: `{{ ssl_router_nginx_config_dir }}{{ custom_host_name }}.cert;`

**Comment by zdmytriv** [comment]
> @zdmytriv if we do this we'll need to make sure that the cert manager is aware of this new directory structure. is there any issue there?

I'm pretty sure that it will be easy to fix.

**Comment by max-lobur** [comment]
@mjallday Except the DNS resolvers update that worked

**Pull request merged**